server/src/core/user/user-service.ts

```diff
@@ -79,6 +87,40 @@ export class Users implements UserService {
     }
   }
 
+  async nominate(args: UserNominateArgs): Promise<User> {
+    const { phone, nominator } = args;
+    try {
+      const user = await this.collection.findOne({ phone });
+
+      if (user && 'beneficiary' in user.roles) {
+        const result = await this.collection.findOneAndUpdate(
+          { phone }, 
+          { $addToSet: { donors: nominator }, $currentDate: { updatedAt: true } },
+          { upsert: true, returnOriginal: false }
+        );
+        return getSafeUser(user);
+      }
+      else if (user && 'middleman' in user.roles) {
+        const result = await this.collection.findOneAndUpdate(
+          { phone }, 
+          { $addToSet: { donors: nominator, roles: 'beneficiary' }, $currentDate: { updatedAt: true } }, 
+          { upsert: true, returnOriginal: false }
+        );
+        return getSafeUser(user);
+      }
+      else if (user && 'donor' in user.roles) {
+        throw createNominateFailedError();
+      }
+      else {
+        return await this.create({ phone, password: '', addedBy: nominator, role: 'beneficiary' });
+      }
```

Comment:
Try to do this in one db operation. Fetching a document in one operation, then checking a field and then updating the document conditionally is not safe. The documentation could be modified in between those two operations. But when you find and update it all in one operation, then it's guaranteed to be atomic.